Proposal for import and dependency system

Running-Lag opened the pull request:
This includes the first version of the import and dependency system.

~~The following things still have to be discussed:~~
~~- Should quotes be required in the as-syntax?~~
        ~~- import "http/client" as http_client vs import "http/client" as "http_client"~~
~~- Should methods/functions inside a struct be callable with the regular function syntax? If yes, how?~~

Dari-OS committed:
Add GitHub actions/workflows (#16)

Running-Lag committed:
First version of the proposal

Running-Lag committed:
Made a few parts of the proposal more clear

Running-Lag commented on docs/IMPORTS-DEPENDENCIES.md:
```diff
@@ -0,0 +1,115 @@
+3. `<Name of an external dependency>`: Relative to an external dependency (described later in this document)
+
+Path relative to origin consists of any directories that need to be traversed and then the file without the file extension.
+This means that module paths can only reference files and never directories.
```
What does @Dari-OS think?

GHaxZ replied:
> I understood that a file is a package...

I think this would be very bad for readability, since ALL string functionality would then have to be written in a single file.

GHaxZ commented on docs/IMPORTS-DEPENDENCIES.md:
```diff
@@ -0,0 +1,115 @@
+### 3.5 Other
+Any project can be an external dependency. While any project with a main function can be run on its own, they too can be depended on.
+This also means that dependencies can have dependencies themselves. Furthermore, dependency cycles are permitted and do not need special
+handling from the user. Multiple dependencies with the same name aren't supported. Dependency versioning isn't supported.
```
Just room for some discussion, should we in any way differentiate between binary and library projects? I think it would be a little weird if libraries were allowed to contain main functions. But this is a very specific discussion.

Running-Lag replied:
I believe that this would just add complexity for no gain.

GHaxZ commented on docs/IMPORTS-DEPENDENCIES.md:
```diff
@@ -0,0 +1,115 @@
+3. `<Name of an external dependency>`: Relative to an external dependency (described later in this document)
+
+Path relative to origin consists of any directories that need to be traversed and then the file without the file extension.
+This means that module paths can only reference files and never directories.
```
I do think we should rather work on a directory package base. This would allow for more flexibility in terms of code organization.

GHaxZ commented on docs/IMPORTS-DEPENDENCIES.md:
```diff
@@ -0,0 +1,115 @@
+Symbol inside imported files are accessed via the following syntax:
+
+`<Location>.<Symbol>`
+By default, location is the filename, but this can be overridden (more on later)
```
Again, I think this should specify the package instead.

Running-Lag replied:
Done

GHaxZ commented on docs/IMPORTS-DEPENDENCIES.md:
```diff
@@ -0,0 +1,115 @@
+### Cyclic imports
+
+Imports may be cyclic without having to follow any special rules
+// TODO: Syntax of symbols inside other symbols (e.g.: structs)
```
As discussed, we should probably disallow struct methods that do not require an instance of the struct. This would eliminate this issue.

GHaxZ reviewed (commented):
Some thoughts and questions from my side

GHaxZ commented:
Additionally, I think we should settle on the `import "http/client" as http_client` syntax. It makes it more clear, that `http_client` is now a symbol for accessing the package.

GHaxZ commented on docs/IMPORTS-DEPENDENCIES.md:
```diff
@@ -0,0 +1,115 @@
+The origin is one of the following: 
+1. `.`: Relative to the current directory.
+2. `<Projectname>`: Relative to the current project
+3. `<Name of an external dependency>`: Relative to an external dependency (described later in this document)
```
Yes, I saw line 96, but it only tells me that importing `import "math/vector"` twice for example is not possible, it is not made that clear that two packages with the same name will cause errors. I just think it could be a bit more clear.

Running-Lag replied:
I made the explanation more clear

Running-Lag committed:
Applied feedback

Running-Lag commented:
> Additionally, I think we should settle on the `import "http/client" as http_client` syntax. It makes it more clear, that `http_client` is now a symbol for accessing the package.

Done

Running-Lag commented:
This should be ready for review

Running-Lag committed:
Update IMPORTS-DEPENDENCIES.md

Fixed a few problems with this proposal

Running-Lag committed:
Merge branch 'main' into docs/import-dependencies

Running-Lag committed:
Fix typos and enhance clarity in IMPORTS-DEPENDENCIES.md

Corrected typos and improved clarity in the documentation.

Running-Lag committed:
Removed unwanted line breaks

LuisNHB commented on docs/IMPORTS-DEPENDENCIES.md:
```diff
@@ -0,0 +1,128 @@
+It may also be empty if there is nothing to be traversed. In this case, the first slash may be omitted.
+
+With the exception of the dot and slash when used as described above, only alphanumerical characters, underscores and hyphens, 
+with the latter two being forbidden at either the start or end of a part of a part, are allowed in `Module Paths`.
```
I think there is a typo. "of a part of a part"

LuisNHB reviewed (commented):
some small suggestions and typos

GHaxZ commented on docs/IMPORTS-DEPENDENCIES.md:
```diff
@@ -0,0 +1,128 @@
+Where Path is the path of the module being imported. It follows the layout of `Module Path` as described above.
+
+For example, the following imports the module `trigonometry` from the math project:
+`import "math/floating_point"`
```
True, this is not the case.

Running-Lag committed:
Apply suggestion

Co-authored-by: Luis <luis.neuhuber@proton.me>

Running-Lag committed:
Apply final round of suggestions

Running-Lag committed:
Removal of redundant escape characters

Hopefully the last change

Running-Lag committed:
Update imports section for clarity and accuracy

Clarified the definition of imports and corrected terminology regarding usage names.
